### PR TITLE
Implement Certificate Revocation Lists in security authentication

### PR DESCRIPTION
**Pull request by clalancette**
This PR implements Certificate Revocation Lists for security authentication in CycloneDDS.  Based on the conversation in #829, this mechanism works by creating a new property QoS named `org.eclipse.cyclonedds.sec.auth.crl` that holds the filename or data for a CRL.  If that property is specified, then the CRL is loaded into the OpenSSL `X509_STORE` before attempting verification.  If the certificate has been revoked, it will be rejected (both by the local side, and by the remote side during handshaking).

I've also implemented testing of this mechanism via `validate_local_identity`.

Note that there are a number of cleanup commits preceding the final two commits that actually implement the feature.  If we want, I could split them into separate PRs as they can go in before this PR.

I have not yet implemented tests for the new CRL mechanism during `begin_handshake_reply` and `process_handshake`.  I'm working on that, but I wanted to get feedback if this is the right direction in the meantime.

I opened this PR against the iceoryx branch, as that is what is currently being used in ROS 2.

FYI @dennis-adlink @eboasson .

### CODE REVIEW

**Review by eboasson — COMMENTED** [review]
Thank you very much @clalancette (and with special thanks for the dusting off of the test cases). I have very few comments, all but one are attached to the changes. The remaining one concerns Cyclone's ability to load the security configuration from the configuration. It may not be used very often, but it does come in quite handy at times. I think it makes sense to add it there, too, but I'm definitely open to arguments otherwise.

That mechanism checks for the presence of any security-related properties in the QoS. If any are given, it must be a complete configuration; if none are given it'll construct a property list QoS from the settings. The key bits are:
* https://github.com/eclipse-cyclonedds/cyclonedds/blob/19fd44e12ad3861944398529469f6711e217bb3f/src/core/ddsi/src/q_entity.c#L903 where it checks for the presence of properties (I think the "org.eclipse.cyclonedds.sec." prefix should be included here, too);
* https://github.com/eclipse-cyclonedds/cyclonedds/blob/19fd44e12ad3861944398529469f6711e217bb3f/src/core/ddsi/src/ddsi_plist.c#L3476 which converts the configuration settings into QoS properties
* https://github.com/eclipse-cyclonedds/cyclonedds/blob/19fd44e12ad3861944398529469f6711e217bb3f/src/core/ddsi/include/dds/ddsi/ddsi_cfgelems.h#L303 where the settings are defined, and
* https://github.com/eclipse-cyclonedds/cyclonedds/blob/19fd44e12ad3861944398529469f6711e217bb3f/src/core/ddsi/include/dds/ddsi/ddsi_config.h#L196 where the type definition is.

The option documentation is generated automatically, but the output of the generator is also committed in the repo so we can easily refer to it. Because of that the CI compares the generated output to the committed one, so you might want to also copy the generated bits to `etc/cyclonedds.rnc`, `etc/cyclonedds.xsd` and `docs/manual/options.md`.

File: src/security/builtin_plugins/authentication/src/authentication.c [hunk 1]
```diff
@@ -1338,14 +1364,14 @@ static X509 *load_X509_certificate_from_binprop (const DDS_Security_BinaryProper
...
-    result = verify_certificate (cert, own_ca, ex);
+    result = verify_certificate (cert, own_ca, own_crl, ex);
   else
   {
     DDS_Security_Exception_clean (ex);
     for (unsigned i = 0; i < trusted_ca_list->length; ++i)
     {
       DDS_Security_Exception_reset (ex);
-      if ((result = verify_certificate (cert, trusted_ca_list->buffer[i], ex)) == DDS_SECURITY_VALIDATION_OK)
+      if ((result = verify_certificate (cert, trusted_ca_list->buffer[i], NULL, ex)) == DDS_SECURITY_VALIDATION_OK)
```
[the remark this answers is not in the document]
Comment (clalancette) [reply src/security/builtin_plugins/authentication/src/authentication.c:1374]: > It seems strange to not apply the CRL if there is a list of trusted CAs, that means that a small change in the way the CA is configured could cause the CRL to be ignored. Perhaps for this it would be more elegant to allow a set of CRLs (like the it allows a set of CAs), but that is much more work for very little gain.

Yeah, we could allow that, but it seemed overkill to me.  There's always room in the future to define a new option for multiple of these if we really want.

> If it is unwise to apply one CRL for different CAs, then I would just fail the check if a CRL is configured and there are multiple CAs. At least that is fail-closed.

As far as I understand, the CRL is signed by a particular CA, and so passing it through to the other set of trusted CAs would result in failures to decrypt it properly.  In that sense, the property is really "crl-from-the-root-CA".  But I like your idea of failing the check if both CRL and trusted_ca_list are both provided.  I'll go ahead and implement that instead.

**Comment by clalancette** [comment]
> That mechanism checks for the presence of any security-related properties in the QoS. If any are given, it must be a complete configuration; if none are given it'll construct a property list QoS from the settings. The key bits are:

Thanks for the review and feedback.  I've gone ahead and implemented this, along with tests for it.

I also made a few bugfixes while working on that.

Finally, I added tests for `begin_handshake`.

That leaves 3 things I know of to do here:

1. Add a test for failure when both trusted_ca_list and CRL are specified.
2. Respond to the comment in https://github.com/eclipse-cyclonedds/cyclonedds/pull/834#discussion_r650763843
3. Add tests for `process_handshake`.

In the meantime, any additional feedback is welcome.

File: src/security/core/tests/config.c [hunk 2]
```diff
@@ -591,6 +589,21 @@ CU_Test(ddssec_config, other_prop, .init = ddsrt_init, .fini = ddsrt_fini)
...
+  dds_qset_prop(qos, DDS_SEC_PROP_CRYPTO_LIBRARY_FINALIZE, "finalize_test_cryptography_all_ok");
+  dds_qset_prop(qos, DDS_SEC_PROP_ACCESS_LIBRARY_PATH, WRAPPERLIB_PATH("dds_security_access_control_wrapper"));
+  dds_qset_prop(qos, DDS_SEC_PROP_ACCESS_LIBRARY_INIT, "init_test_access_control_all_ok");
+  dds_qset_prop(qos, DDS_SEC_PROP_ACCESS_LIBRARY_FINALIZE, "finalize_test_access_control_all_ok");
+  dds_qset_prop(qos, "dds.sec.auth.identity_ca", TEST_IDENTITY_CA_CERTIFICATE_DUMMY);
+  dds_qset_prop(qos, "dds.sec.auth.private_key", TEST_IDENTITY_PRIVATE_KEY_DUMMY);
+  dds_qset_prop(qos, "dds.sec.auth.identity_certificate", TEST_IDENTITY_CERTIFICATE_DUMMY);
+  dds_qset_prop(qos, "dds.sec.access.permissions_ca", "file:Permissions_CA.pem");
+  dds_qset_prop(qos, "dds.sec.access.governance", "file:Governance.p7s");
+  dds_qset_prop(qos, "dds.sec.access.permissions", "file:Permissions.p7s");
```
Comment (dennis-adlink) [inline src/security/core/tests/config.c:613]: @clalancette, the idea behind this test is that security settings from the (xml) configuration are used only when no QoS properties exist that have a name that starts with `dds.sec.`. If any property with this prefix is set, then the security settings from the xml configuration are ignored and all required settings should be set in the QoS. The test case checks this behavior. 

The comment above the test case is not very clear. The terms 'config' and 'config security elements' are confusing and should be something like '_Cyclone XML configuration_' and '_security element in the Cyclone XML configuration_'. 

I think the check 'any security properties in the QoS' should also include the vendor specific prefix that is introduced by this PR.

Comment (clalancette) [reply src/security/core/tests/config.c:613]: Hm, OK.  The reason I made this change at all is that this test was failing to load the security modules, which ended up in a crash.  I honestly don't understand the module loading mechanism at present, so I wasn't sure what was going on here.  I'll take another look at this test and see if I can figure out the right thing to do.

File: src/security/builtin_plugins/tests/validate_local_identity/src/validate_local_identity_utests.c [hunk 3]
```diff
@@ -2141,13 +2265,96 @@ CU_Test(ddssec_builtin_validate_local_identity,with_extended_certificate_check)
...
+{
+    DDS_Security_ValidationResult_t result;
+    DDS_Security_IdentityHandle local_identity_handle = DDS_SECURITY_HANDLE_NIL;
+    DDS_Security_GUID_t adjusted_participant_guid;
+    DDS_Security_DomainId domain_id = 0;
+    DDS_Security_Qos participant_qos;
+    DDS_Security_GUID_t candidate_participant_guid;
+    DDS_Security_SecurityException exception = {NULL, 0, 0};
+    DDS_Security_GuidPrefix_t prefix = {0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8, 0xf9, 0xfa, 0xfb};
+    DDS_Security_EntityId_t entityId = {{0xa0,0xa1,0xa2},0x1};
```
[the remark this answers is not in the document]
Comment (clalancette) [reply src/security/builtin_plugins/tests/validate_local_identity/src/validate_local_identity_utests.c:2248]: So I looked at this one briefly, and I'm not sure I understand where the `0xc1` is supposed to be.  All of the places in this file where `print_guid` is called end up with something like:

```
303: adjusted_participant_guid=f767762b:c1ef319a:70a383fa:a0a1a2:01
303: adjusted_participant_guid=f767762b:c1ef319a:70a383fa:a0a1a2:01
```

Where the `entityKind` member is `0x01`.  That goes both for tests that succeed and ones that fail.  Could you explain a little more about what you are looking for here?

[the remark this answers is not in the document]
Comment (eboasson) [reply src/security/builtin_plugins/tests/validate_local_identity/src/validate_local_identity_utests.c:2248]: You are entirely correct, I simply hadn't realised that all of those tests were using GUIDs with `entityKind` 1. By my reasoning that would make them all "wrong" but it doesn't make any sense to extend this PR to include that matter.

Anyway, in my reading of the DDSI spec, the only possible entity kind for participant GUIDs allowed is `0xc1` , and so I check this for any incoming participant GUID and reject any discovery message for which it doesn't hold:
https://github.com/eclipse-cyclonedds/cyclonedds/blob/5fa204dd61620b24c394ef88d9040944ecbbb470/src/core/ddsi/src/ddsi_plist.c#L1608-L1616

With that in mind, it is quite possible that there is no value whatsoever in having `validate_xyz_identity`, or indeed, any of the security plugins, pay attention to the value of `entityKind` as long as it simply copies the `entityId`/`entityKind` from its input as spec'd. (Which they do here.)

Still, there are some GUIDs sent in, e.g., handshake messages and so I should check if those are all validated properly, or else if a bogus `entityId`/`entityKind` can do no harm. For example, table 52 of the Security spec says:

> The operation shall verify that the first bit of the participant_guid of the ParticipantBuiltinTopic data inside the “c.pdata” is set to 1 and that the following 47 bits match the first 47 bits of the SHA-256 hash of the SubjectName appearing in the IdentityCredential. If this verification fails, the operation shall fail and return ValidationResult_Fail.

According to the spec then, the `entityId`/`entityKind` should be ignored in the validation, and that means I need evidence that there is no harm in a bogus `entityId`/`entityKind`. But all of that is outside the scope of this PR!

File: src/security/core/tests/config.c (hunk 2 above)
Comment (clalancette) [reply src/security/core/tests/config.c]: This turned out to be a bug in ddsi_plist.c; see 38f1e56 .  I've now restored this code to the original (plus the small changes needed for the new property).

File: src/security/builtin_plugins/tests/validate_local_identity/src/validate_local_identity_utests.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (clalancette) [reply src/security/builtin_plugins/tests/validate_local_identity/src/validate_local_identity_utests.c:2248]: Thanks for the analysis, makes more sense now.  If you'd like, I can open an issue so someone can look at this later on.  In the meantime, I'll mark this conversation as "Resolved" since there isn't anything to do for this PR.

**Comment by clalancette** [comment]
All right, the latest set of commits should fix all of the outstanding items, including adding a test for process_handshake with CRLs.  Assuming CI is green, I think this is ready for another round of review.

**Review by eboasson — COMMENTED** [review]
I don't much in the way of comments, I'd rather classify it as a remark, a suggestion and one comment:
* The remark is that I agree with your decision to perpetuate the eyesore that is all those string literals with property names sprinkled throughout. I have always wanted to do something about that, and now your discovery of the misspelled `dds.ca.algo` has pushed me to do it. There'll be a PR from me soon (I think I'm about 80% done) that result in a single set of definitions that are then used everywhere. It has absolutely no bearing on this PR, but I thought it only proper to thank you for this.
* The suggestion is that it is probably worth considering what the best behaviour is when the XML specifies a CRL while the participant QoS provides a full security configuration without a CRL. It is a perfectly legitimate configuration, but it could be argued that it would better if the QoS had to explicitly set the CRL in this case. I am really of two minds on this. I gave it a shot, so if you would be willing to give this some consideration, there is an implementation at https://github.com/eboasson/cyclonedds/tree/pr-834b
* The comment is really just a commit on that branch https://github.com/eboasson/cyclonedds/commit/8a3bb3875777035f095fdd957a4f88b9b8eb866e — @dennis-adlink commented along those lines and I think he is right.

The CI is green, so apart from giving the suggestion and comment some consideration, it is good to go. Thank you!

**Comment by clalancette** [comment]
> * The comment is really just a commit on that branch [eboasson@8a3bb38](https://github.com/eboasson/cyclonedds/commit/8a3bb3875777035f095fdd957a4f88b9b8eb866e) — @dennis-adlink commented along those lines and I think he is right.

Yeah, so CRLs are actually in a weird position here.  With the exception of `trusted_ca_dir`, the failure to specify any of the `dds.sec` properties results in failure to start the participant, which keeps things secure.  `trusted_ca_dir` is a little different in that it is optional, but it also an allow list; if you don't specify the additional CAs that you need, you may fail to handshake the remote node.  This is also secure.

CRLs are different.  They deny entry to an otherwise authorized participant, but they are also optional.  So if you fail to specify the CRL properly, you'll silently allow participants to connect that should not be allowed anymore.

From a usability standpoint, however, CRLs don't seem to be very common.  And the problem with incorrectly specifying a CRL isn't just in CycloneDDS; the same conceptual problem exists in the underlying OpenSSL library.  Ultimately I decided for practicality over security here.  That is, since CRLs aren't that common, I decided to keep the DDS property optional.  We *could* make it required (as you do in https://github.com/eboasson/cyclonedds/commit/8a3bb3875777035f095fdd957a4f88b9b8eb866e), but it means that most users will need to present an "empty" CRL there (including ROS 2).  It also means that `rmw_cyclonedds` would deviate from both `rmw_fastrtps` and `rmw_connextdds` in this respect.

At the end of the day, we can accommodate that (both in CycloneDDS and in ROS 2), but it just makes it even harder for users to use security.  Let me know how you feel about it; if you still feel we should make it a required property, I'll go ahead and cherry-pick your commits over to this PR.

**Comment by eboasson** [comment]
> At the end of the day, we can accommodate that (both in CycloneDDS and in ROS 2), but it just makes it even harder for users to use security. Let me know how you feel about it; if you still feel we should make it a required property, I'll go ahead and cherry-pick your commits over to this PR.

I don't think CRLs are common enough to warrant making life harder for people. But the intent behind the code that my commit changes a bit is that it says: if *some* security property is set, then there is a set of properties that must have been provided. It doesn't literally require all properties to be present. E.g., if you set `org.eclipse.cyclonedds.sec.auth.crl` in the QoS, then the assumption is that you intended to enable security and we'd better verify all properties needed for that are present.

Neither `trusted_ca_dir` nor the CRL is required for a meaningful security configuration, and therefore neither is in the list of properties that it checks given the intent to enable security (i.e., https://github.com/eclipse-cyclonedds/cyclonedds/blob/e34d38af35713297f0865a32f8889dca925e2d8c/src/core/ddsi/src/q_entity.c#L913).

So I don't see how that particular commit would make life harder, CRLs would still be as optional as they ever were and not specifying one remains perfectly acceptable.

In short, maybe I am missing something. If adding that commit makes life more complicated, then I am in favour of merging your PR as it is, else I do would like to merge it.

(Regarding the other additions on that branch (the ones that say: "if the XML configuration is specified a CRL, then you must override it explicitly"): I wasn't sure about it to begin with because it tries to infer intent on very weak grounds. Too much nanny-state-ism in there, in hindsight. It isn't often that hindsight comes in time 🙂.)

**Comment by clalancette** [comment]
> So I don't see how that particular commit would make life harder, CRLs would still be as optional as they ever were and not specifying one remains perfectly acceptable.

Oh, I see.  Yeah, you are correct.  Sorry, I misread that commit entirely!  In that case, things look good to me, and I've now cherry-picked those commits off of your branch and into here.  Please take another look when you get a chance.

**Review by eboasson — APPROVED** [review]
LGTM! Thanks @clalancette

**Comment by eboasson** [comment]
@k0ekk0ek this currently targets the `iceoryx` branch but we of course want it on `master`. I do think it makes sense to also make it available in the `iceoryx` branch as that is what ROS 2 is currently using. As I would like to avoid any further merges between the two branches, I propose to simply cherry-pick them into `master`. Do you agree?

**Comment by k0ekk0ek** [comment]
Sounds like a plan :+1:

**Comment by eboasson** [comment]
Sanity check with a limited ROS 2 CI run:

* Linux x64 [Build Status](https://ci.ros2.org/job/ci_linux/14804/)
* Linux aarch64 [Build Status](https://ci.ros2.org/job/ci_linux-aarch64/9553/)
* macOS [Build Status](https://ci.ros2.org/job/ci_osx/12470/)
* Windows [Build Status](https://ci.ros2.org/job/ci_windows/14957/)

**Comment by clalancette** [comment]
Thanks!
